File: lib/gptzero.ts
import type { GptZeroFlag } from "./types";

export interface GptZeroResult {
  score: number | null;
  flag: GptZeroFlag;
}

export async function checkGptZero(text: string): Promise<GptZeroResult> {
  const apiKey = process.env.GPTZERO_API_KEY;
  if (!apiKey) return { score: null, flag: "unavailable" };

  try {
    const res = await fetch("https://api.gptzero.me/v2/predict/text", {
      method: "POST",
      headers: {
        "x-api-key": apiKey,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ document: text }),
      signal: AbortSignal.timeout(8000),
    });

    if (!res.ok) return { score: null, flag: "unavailable" };

    const data = await res.json();
    const doc = data?.documents?.[0];
    if (!doc) return { score: null, flag: "unavailable" };

    const probs = doc.class_probabilities as
      | { human?: number; mixed?: number; ai?: number }
      | undefined;
    const score: number = doc.completely_generated_prob ?? probs?.ai ?? 0;

    let flag: GptZeroFlag = "human";
    if (probs) {
      const entries = Object.entries(probs) as [
        "human" | "mixed" | "ai",
        number
      ][];
      flag = entries.reduce((a, b) => (b[1] > a[1] ? b : a))[0];
    } else {
      flag = score > 0.5 ? "ai" : "human";
    }

    return { score, flag };
  } catch {
    return { score: null, flag: "unavailable" };
  }
}

File: backend/integrations/gptzero.py
"""GPTZero AI-authorship detection.

POST https://api.gptzero.me/v2/predict/text with an `x-api-key` header.
Returns `{"ai_probability": float, "flagged": bool}` — the shape the Verdict wants.
"""

from __future__ import annotations

import os

import httpx

from . import TIMEOUT, expected_for, log, safe_call

API_URL = "https://api.gptzero.me/v2/predict/text"

#: Above this, the report never auto-approves. Contract non-negotiable #2.
FLAG_THRESHOLD = 0.85


def _extract_probability(payload: dict) -> float:
    """Pull the AI probability out of a v2 response, tolerating field drift."""
    docs = payload.get("documents") or []
    doc = docs[0] if docs else payload
    for key in ("completely_generated_prob", "average_generated_prob"):
        value = doc.get(key)
        if isinstance(value, (int, float)):
            return float(value)
    ai = (doc.get("class_probabilities") or {}).get("ai")
    if isinstance(ai, (int, float)):
        return float(ai)
    # Last resort: a categorical verdict with no score attached.
    return 1.0 if doc.get("predicted_class") == "ai" else 0.0


async def score_text(text: str | None, task_id: str) -> dict:
    """Score `text` for AI authorship. Always returns a well-formed dict."""
    fallback = dict(expected_for(task_id).get("gptzero") or {"ai_probability": 0.05, "flagged": False})

    if not text or not text.strip():
        # Nothing written by a human or a machine — voice-only submissions land here.
        return {"ai_probability": 0.0, "flagged": False}

    key = os.getenv("GPTZERO_API_KEY")
    if not key:
        log.warning("gptzero: GPTZERO_API_KEY not set, using canned fallback")
        return fallback

    async def call() -> dict:
        async with httpx.AsyncClient(timeout=TIMEOUT) as client:
            resp = await client.post(
                API_URL,
                headers={"x-api-key": key, "Content-Type": "application/json"},
                json={"document": text},
            )
            resp.raise_for_status()
            prob = _extract_probability(resp.json())
        return {"ai_probability": round(prob, 3), "flagged": prob > FLAG_THRESHOLD}

    return await safe_call("gptzero", call, fallback)
